### ISSUE TITLE
Ignore generated code

### ISSUE THREAD

**Issue by stepancheg** [Manishearth/rust-clippy#702]
Rust-protobuf generates code with several clippy warnings. These warnings are hard to fix, they [bother rust-protobuf users](https://github.com/stepancheg/rust-protobuf/issues/133), and they actually do need to be fixed (because I know the code is correct).

So the best way to fix the issue in rust-protobuf with clippy would be probably to tell clippy to ignore all clippy warning in some file. However as I understand there is no way to do that. 

Clippy could ignore files with comments like "this file is generated" or "ignore clippy".

Note that generated code must be valid even if clippy is not used.

**Comment by Manishearth** [comment]
> However as I understand there is no way to do that. 

You can just stick `#![allow(clippy)]` at the top of the module. There's no need for a magic comments that clippy has to read; Rust's lint infrastructure already has this ability.

**Comment by stepancheg** [comment]
`#![allow(unknown_lints)]` hides errors, for example, if rust removes some lint in future version. Otherwise it works. Thanks.

**Comment by crusty-dave** [comment]
I would expect to be able to do the following in scope:

#![warn(clippy::needless_range_loop, clippy::manual_memcpy)]

or possibly

#[!warn(clippy::needless_range_loop, clippy::manual_memcpy)]

In my case it is reporting a false positive that I would like to suppress, but I would like to use it to check other parts of the source code.

**Comment by oli-obk** [comment]
> In my case it is reporting a false positive that I would like to suppress, but I would like to use it to check other parts of the source code.

You can simply attach `#[allow(clippy::foo)]` to the item that causes the false positive. Also I suggest you open an issue about the false positive

**Comment by jkozlowski** [comment]
This all works until I add '-- -D warnings' to the clippy invocation and it upgrades a warning from the compiler and the build. Here's the example:

```
error: hidden lifetime parameters in types are deprecated
   --> /home/jakubkozlowski/Programming/starfish/target/debug/build/spdk-sys-f36c12b8dbef03e3/out/spdk_bindings.rs:127:29
    |
127 |     fn fmt(&self, fmt: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
    |                             ^^^^^^^^^^^^^^^^^^^^^- help: indicate the anonymous lifetime: `<'_>`
```

I have:
```
#![allow(warnings)]
#![allow(clippy)]
#![allow(unknown_lints)]
```

Before the include, does anyone know if it's possible to turn off everything for this file, so I can keep -D warnings?

**Comment by jamesray1** [comment]
> You can just stick #![allow(clippy)] at the top of the module. There's no need for a magic comments that clippy has to read; Rust's lint infrastructure already has this ability.

I get the following compiler info with this line from the generated code:
```
lint name `clippy` is deprecated and may not have an effect in the future. Also `cfg_attr(cargo-clippy)` won't be necessary anymore

note: #[warn(renamed_and_removed_lints)] on by default
help: change it to: clippy::all rustc(renamed_and_removed_lints)

lint name `clippy` is deprecated and may not have an effect in the future. Also `cfg_attr(cargo-clippy)` won't be necessary anymore

help: change it to: clippy::all rustc(renamed_and_removed_lints)
```

Changing to `#![allow(clippy::all)]` fixes this.

**Comment by CreepySkeleton** [comment]
Could anyone please explain why we should ever care about code originates from macro expansions? Such code tends to be redundant, complex, and unclear - but soulless machine does not and should not care about it.

**Comment by Manishearth** [comment]
Yes, that's what `allow(clippy::all)` is for. Just use that. Detecting if a file is autogenerated is a can of worms we do not want to open.

**Comment by CreepySkeleton** [comment]
@Manishearth I'm not talking about auto-generated files, it's about macro expansions? Any chance we may have it?

**Comment by Manishearth** [comment]
We ignore macro expansions for most lints. File a bug for a specific lint and code example, we can fix it. This is the wrong place for this complaint.

**Comment by Rantanen** [comment]
`allow(clippy::all)` works for proc macro expansion as well. Clippy ignores that expansion, until the proc macro goes out of its way and starts expanding using spans from the user code.

I'm currently working for better error messages for my proc macro attributes. My approach is to use spans of user code when expanding the proc macro attribute so that the rustc error messages would point to the bad bits in the user code instead of the `#[com_interface]` attribute.

This resulted in clippy not ignoring those bits anymore. Google brought me here and `allow(clippy::all)` provided the solution. I'm now adding that to all the new items that the attribute is expanding to.

(Sorry for the noise, figured I'd clarify this here given this issue was rather high up in Google results.)

**Comment by CreepySkeleton** [comment]
@Rantanen yes, this is exactly what [we did in the end](https://github.com/TeXitoi/structopt/pull/263)

**Comment by Timmmm** [comment]
Uhm, is this supposed to work?

```
warning: unknown lint: `clippy`
 --> src/capnp/mod.rs:1:36
  |
1 | #[allow(non_snake_case, dead_code, clippy)]
  |                                    ^^^^^^
  |
  = note: `#[warn(unknown_lints)]` on by default
```

Using stable Clippy as of today.

**Comment by Timmmm** [comment]
Ah right, I see - you have to put `#![allow(unknown_lints)]` in the file where you do that. A bit of a shame Rust doesn't know about `clippy` by default, given that it is now included in `rustup`.

**Comment by Manishearth** [comment]
@Timmmm no, you need to do `#![allow(clippy::all)]`. Rustc does know about clippy, but you need to use the namespace

**Comment by AsadullahFarooqi** [comment]
> > However as I understand there is no way to do that.
> 
> You can just stick `#![allow(clippy)]` at the top of the module. There's no need for a magic comments that clippy has to read; Rust's lint infrastructure already has this ability.
> 
> You may also need to `#![allow(unknown_lints)]` on the same file.



> Yes, that's what `allow(clippy::all)` is for. Just use that. Detecting if a file is autogenerated is a can of worms we do not want to open.

I'm trying to disable clippy on auto-generated protobuf (tokio-rs/prost) but not sure what to do, I've tried the above suggestions but it still gives warnings on auto-generated files. I've tried `#allow(clippy::all)` the builder.rs 
https://github.com/AsadullahFarooqi/lance/tree/main/rust

**Comment by llogiq** [comment]
There is no builder.rs file in that repo. Are you sure you have the `#[allow(clippy::all)]` in the generated code or in a module above?

**Comment by AsadullahFarooqi** [comment]
here is the builder.rs link
https://github.com/eto-ai/lance/blob/main/rust/build.rs
the generated code is used in the `src/index.rs` and `src/format.rc`

**Comment by Manishearth** [comment]
You need to either

 - have `#![allow(clippy::all)]` (note the exclamation mark) at the top of the _generated_ file
 - have `#[allow(clippy::all)]` (no exclamation mark) on the `mod` statement pointing to the generated code


We can't really help you without being able to see the generated code or how it's pulled in

**Comment by sliedes** [comment]
> If `#[allow(clippy::all)]` alone does not work for you, try the following:
> 
> ```
> #[allow(clippy::all, clippy::pedantic, clippy::restriction, clippy::nursery)]
> pub mod hello {
>     tonic::include_proto!("hello");
> }
> ```

This helped for me. Ok, I'm quite confused about `clippy::all` not being sufficient? Sounds also mildly pathological if we need to rely on lint categories not being added later...
